### PR TITLE
chore(core): enabled websocket in route

### LINKED ISSUE #1791: change: enabled websocket  in route
I think websocket should be enabled in route, not upstream. This logic is clearer. On the other hand, if we support multiple upstream, we only need to configure it once in the route, otherwise we need to configure each upstream.

### PR DESCRIPTION
**Pull request by dabue**
change: enabled websocket in route 

resolve #1791

### Pre-submission checklist:

* enable websocket 
* update the test cases
* update the user's guide
* **this is a non-backward compatible PR**

### CODE REVIEW

**Comment by membphis** [comment]
@dabue I think we need some test case about the `enable_websocket` was used in `service` object.

For this changing:

https://github.com/apache/apisix/pull/2000/files#diff-d982d52466e7c93c7b604358339b2a29R378-R380

**Comment by membphis** [comment]
> Is this a non-backward compatible PR?

yes

**Comment by dabue** [comment]
ok, I'll update it.



88786912@qq.com
 
From: YuanSheng Wang
Date: 2020-08-05 22:37
To: apache/apisix
CC: dabue; Mention
Subject: Re: [apache/apisix] [change] enabled websocket in route (#2000)
@dabue I think we need some test case about the enable_websocket was used in service object.
For this changing:
https://github.com/apache/apisix/pull/2000/files#diff-d982d52466e7c93c7b604358339b2a29R378-R380

**Comment by dabue** [comment]
I have updated it , please help to check it, thx.



88786912@qq.com
 
From: dabue
Date: 2020-08-09 10:03
To: apache/apisix; apache/apisix
CC: mention
Subject: Re: Re: [apache/apisix] [change] enabled websocket in route (#2000)
ok, I'll update it.



88786912@qq.com
 
From: YuanSheng Wang
Date: 2020-08-05 22:37
To: apache/apisix
CC: dabue; Mention
Subject: Re: [apache/apisix] [change] enabled websocket in route (#2000)
@dabue I think we need some test case about the enable_websocket was used in service object.
For this changing:
https://github.com/apache/apisix/pull/2000/files#diff-d982d52466e7c93c7b604358339b2a29R378-R380

File: apisix/schema_def.lua [hunk 1]
```diff
@@ -465,6 +465,10 @@ _M.service = {
         name = {type = "string", maxLength = 50},
         desc = {type = "string", maxLength = 256},
         script = {type = "string", minLength = 10, maxLength = 102400},
+        enable_websocket = {
```
Comment (membphis) [inline apisix/schema_def.lua:468]: we need to update the doc about `service`

**Comment by moonming** [comment]
why not supports enable websocket in both route and service?

YuanSheng Wang <notifications@github.com> 于 2020年8月7日周五 下午9:55写道：

**Comment by moonming** [comment]
@dabue ping

**Comment by membphis** [comment]
> why not supports enable websocket in both route and service?

That is good if we want to enable `enable_websocket` in the `service` level.

File: apisix/init.lua [hunk 2]
```diff
@@ -442,15 +442,12 @@ function _M.http_access_phase()
...
-
-        if route.value.upstream and route.value.upstream.enable_websocket then
-            enable_websocket = true
-        end
     end
 
     if enable_websocket then
         api_ctx.var.upstream_upgrade    = api_ctx.var.http_upgrade
         api_ctx.var.upstream_connection = api_ctx.var.http_connection
+        core.log.info("enabled websocket for route: ", route.value.id)
```
Comment (membphis) [inline apisix/init.lua:450]: why we need this?

Comment (membphis) [reply apisix/init.lua:450]: wow, I got it, for testing. ignore what I said.

**Comment by dabue** [comment]
> @dabue ping

This function has been supported in route and service. I‘ll update the  docs.

File: apisix/schema_def.lua (hunk 1 above)
Comment (dabue) [reply apisix/schema_def.lua:468]: I have updated the admin-api.md, please check it.

**Comment by membphis** [comment]
@dabue pls rebase your branch

**Comment by membphis** [comment]
@dabue please rebase your branch, then we can continue this PR ^_^

File: README.md [hunk 3]
```diff
@@ -91,6 +91,7 @@ A/B testing, canary release, blue-green deployment, limit rate, defense against
   - [SSL](doc/https.md): Dynamically load an SSL certificate.
 
 - **Full dynamic**
+<<<<<<< HEAD
```
Comment (juzhiyuan) [inline README.md:94]: Ping @dabue, something error happened when you are resolving conflicts.

File: README.md [hunk 4]
```diff
@@ -91,7 +91,6 @@ A/B testing, canary release, blue-green deployment, limit rate, defense against
   - [SSL](doc/https.md): Dynamically load an SSL certificate.
 
 - **Full dynamic**
-
```
Comment (membphis) [inline README.md:94]: one PR for one thing, you can do this in a new PR.

File: doc/admin-api.md [hunk 5]
```diff
@@ -70,7 +70,11 @@
 |upstream_id|False |Upstream|Enabled upstream id, see [Upstream](architecture-design.md#upstream) for more ||
 |service_id|False |Service|Binded Service configuration, see [Service](architecture-design.md#service) for more ||
 |service_protocol|False|Upstream protocol type|only `grpc` and `http` are supported|`http` is the default value; Must set `grpc` if using `gRPC proxy` or `gRPC transcode`|
+<<<<<<< HEAD
```
Comment (membphis) [inline doc/admin-api.md:73]: it seems that you did not rebase your branch in the correct way.

please fix it.

File: doc/admin-api.md [hunk 6]
```diff
@@ -289,7 +295,11 @@ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f13
 |upstream_id|False |Upstream|Enabled upstream id, see [Upstream](architecture-design.md#upstream) for more ||
 |name     |False |Auxiliary   |Identifies service names.|customer-xxxx|
 |desc     |False |Auxiliary   |service usage scenarios, and more.|customer xxxx|
+<<<<<<< HEAD
```
Comment (membphis) [inline doc/admin-api.md:298]: ditto

File: doc/architecture-design.md [hunk 7]
```diff
@@ -262,9 +262,9 @@ In addition to the basic complex equalization algorithm selection, APISIX's Upst
...
 |checks          |optional|Configure the parameters of the health check. For details, refer to [health-check](health-check.md).|
 |retries         |optional|Pass the request to the next upstream using the underlying Nginx retry mechanism, the retry mechanism is enabled by default and set the number of retries according to the number of backend nodes. If `retries` option is explicitly set, it will override the default value.|
-|enable_websocket|optional| enable `websocket`(boolean), default `false`.|
 |timeout|optional| Set the timeout for connection, sending and receiving messages. |
 |desc     |optional|Identifies route names, usage scenarios, and more.|
+|enable_websocket|optional| enable `websocket`(boolean), default `false`.|
```
Comment (membphis) [inline doc/architecture-design.md:267]: I think we should remove `enable_websocket` in the upstream object.

please check all similar points.

**Review by membphis — CHANGES_REQUESTED** [review]

**Comment by dabue** [comment]
I will close this pr and open a new pr to resolve the conflict.

**Comment by dabue** [comment]
new pr #2593

**Pull request closed without merging**